INTEGRITY: Update checksum calculation logic for macfiles

ShivangNagta opened the pull request:
* Checksum calculations for data forks and data section of the resource forks have been added for 7 macfile variants - macbinary with .bin, macbinary without .bin, appledouble with .rsrc, appledouble with ._, appledouble with __MACOSX, raw rsrc and mac's actual resource fork.
* File filtering has been added to make sure only a single file entry is added in case of macfiles, e.g for appledouble with <._appledouble_file>(contains resource fork) and <appledouble_file>(contains data fork) , the second file i.e. <appledouble_file> is removed from the checksum calculation dictionary. All the checksums can be calculated using <._appledouble_file>'s filepath.
* All the files are marked under one of the following category - NON_MAC, MAC_BINARY (both .bin and without .bin are handled together), APPLE_DOUBLE_RSRC, APPLE_DOUBLE_MACOSX, APPLE_DOUBLE_DOT_, RAW_RSRC and ACTUAL_FORK_MAC.

ShivangNagta committed:
INTEGRITY: Add file filtering and update checksum calculation logic for 6 mac file variants

* Checksum calculations for data forks and data section of the resource forks have been added for 6 macfile variants - macbinary with .bin, macbinary without .bin, appledouble with .rsrc, appledouble with ._, appledouble with __MACOSX a
nd raw rsrc.
* File filtering has been added to make sure only a single file entry is added in case of macfiles, e.g for appledouble with <._appledouble_file>(contains resource fork) and <appledouble_file>(contains data fork) , the second file i.e.
<appledouble_file> is removed from the checksum calculation dictionary. All the checksums can be calculated using <._appledouble_file>'s filepath.
* All the files are marked under one of the following category - NON_MAC, MAC_BINARY (both .bin and without .bin are handled together), APPLE_DOUBLE_RSRC, APPLE_DOUBLE_MACOSX, APPLE_DOUBLE_DOT_ and RAW_RSRC

ShivangNagta committed:
INTEGRITY: Remove punycode encoding from compute_hash.py

Punycode encoding is already handled by dat parser, i.e during uploading of dat files to the database.

ShivangNagta committed:
INTEGRITY: Add tests for macbinary

ShivangNagta committed:
INTEGRITY: Extend file table with two more size fields

ShivangNagta committed:
INTEGRITY: Changes in web app

* Add size-r and size-rd fields in files
* Add a Clear Database button, for development phase

ShivangNagta committed:
INTEGRTIY: Initial seeding and set.dat fixes.

* Sort files by name before megakey calculation.
* Correctly merging matched set.dat to original detection fileset.
* Conditional size field check as some dats might have three sizes.

ShivangNagta committed:
INTEGRITY: Identify size type from checksum prefix

ShivangNagta committed:
INTEGRITY: Fix fileset deletion logic when matching set filesets with detection.

Previously, the logic attempted to fetch the current fileset ID using the last insert ID, which was incorrect since the last insert was into the `filechecksum` table, not `fileset`. Now, the correct fileset ID is returned directly from the `insert_fileset` function.

ShivangNagta committed:
INTEGRITY: Correct fileset redirection link in log in case of matching

ShivangNagta committed:
INTEGRITY: Add all md5 variants for files with size less than 5000B.

If a file has a size less than 5000, then its checksum is equal for all variants, md5-full_size, md5-5000, md5-t-5000 and md5-1M.

rvanlaar commented on compute_hash.py:
```diff
@@ -192,29 +232,45 @@ def macbin_get_datafork(file_byte_stream):
 def is_appledouble(file_byte_stream):
-    # Check AppleDouble magic number
+    """
+    Appledouble Structure -
```
Nice job documenting the apple double structure.

rvanlaar commented on compute_hash.py:
```diff
@@ -223,48 +279,56 @@ def appledouble_get_datafork(filepath):
+            return (create_checksum_pairs(checksum(f, alg, size, filepath), alg, size), filesize(filepath))
+        
+        # Processing mac files
+        else:
```
No need for else when the part before contains a return. 
Makes for cleaner code with less indentations.
And check if you can eliminate duplicate code in this function as well. What would your options be?

ShivangNagta replied:
Yes, I am using the seek operation unnecessarily. Let me refactor it.

rvanlaar commented on compute_hash.py:
```diff
@@ -223,48 +279,56 @@ def appledouble_get_datafork(filepath):
+                (resfork, size) = macbin_get_resfork_data(f.read())
+                f.seek(0)
+                datafork = macbin_get_datafork(f.read())
+            elif file_info[0] == FileType.APPLE_DOUBLE_DOT_ or file_info[0] == FileType.APPLE_DOUBLE_RSRC or file_info[0] == FileType.APPLE_DOUBLE_MACOSX:
```
Use the `in` keyword instead of 3 different equality checks.

rvanlaar commented on compute_hash.py:
```diff
@@ -379,6 +437,94 @@ def checksum(file, alg, size, filepath):
+        return [FileType.MAC_BINARY, extract_macbin_filename_from_header(filepath)] 
+    
+    # 2. Appledouble .rsrc
+    elif is_appledouble_rsrc(filepath):
```
Because the above statement returns, there's no need to use elif. Change it to if.

rvanlaar commented on compute_hash.py:
```diff
@@ -379,6 +437,94 @@ def checksum(file, alg, size, filepath):
+            expected_data_fork_path = os.path.join(parent_dir_path, file_info[1])
+            if (expected_data_fork_path in files):
+                to_be_deleted.append(expected_data_fork_path)
+            # else:
```
Commented out code blocks are fine when developing, just make sure they don't end up in commits.
Check out how to use `git add -p`.

rvanlaar commented on compute_hash.py:
```diff
@@ -415,8 +580,7 @@ def create_dat_file(hash_of_dirs, path, checksum_size=0):
             for filename, (hashes, filesize) in hash_of_dir.items():
-                filename = (punyencode_filepath(filename)
-                            if needs_punyencoding(filename) else filename)
+                filename = (punyencode_filepath(filename) if needs_punyencoding(filename) else filename)
```
There's no code change here, so it shouldn't belong in this commit.

rvanlaar commented on compute_hash.py:
```diff
@@ -207,8 +208,18 @@ def is_macbin(filepath):
+    # parent_dir_path = os.path.dirname(filepath)
+    resource_fork_path = os.path.join(filepath, "..namedfork", "rsrc")
+    print(resource_fork_path)
+    if os.path.exists(resource_fork_path):
```
This can and should be rewritten as `return os.path.exists(resource_fork_path)`. 
Can you see why?

rvanlaar commented on compute_hash.py:
```diff
@@ -207,8 +208,18 @@ def is_macbin(filepath):
 
+def is_actual_resource_fork_mac(filepath):
+    """ Returns boolean, checking the actual mac fork if it exists. """
+    # filename = os.path.basename(filepath)
```
Make sure commented out code doesn't end up in commits.

rvanlaar commented on compute_hash.py:
```diff
@@ -218,6 +218,26 @@ def is_actual_resource_fork_mac(filepath):
         return True
     return False
 
+def is_appledouble(file_byte_stream):
```
This code move doesn't belong in a commit for "Additional error handling ... "

rvanlaar commented on compute_hash.py:
```diff
@@ -301,8 +301,11 @@ def appledouble_get_datafork(filepath, fileinfo):
+    try:
+        with open(filepath[:-5]+".data", "rb") as f:
+            return f.read()
+    except (FileNotFoundError, IsADirectoryError):
```
good job

rvanlaar commented on compute_hash.py:
```diff
@@ -477,7 +483,7 @@ def extract_macbin_filename_from_header(file):
 
 def file_classification(filepath):
-    """ Returns [ Filetype, Filename ]. Filetype is an enum value - NON_MAC, MAC_BINARY, APPLE_DOUBLE_RSRC, APPLE_DOUBLE_MACOSX, APPLE_DOUBLE_DOT_, RAW_RSRC
+    """ Returns [ Filetype, Filename ]. Filetype is an enum value - NON_MAC, MAC_BINARY, APPLE_DOUBLE_RSRC, APPLE_DOUBLE_MACOSX, APPLE_DOUBLE_DOT_, RAW_RSRC and ACTUAL_FORK_MAC.
```
I applaud the documentation. It should belong in the previous commit.

rvanlaar commented on db_functions.py:
```diff
@@ -160,7 +160,7 @@ def insert_file(file, detection, src, conn):
     detection_type = f"{checktype}-{checksize}" if checktype != "None" else f"{checktype}"
     if punycode_need_encode(file['name']):
-        query = f"INSERT INTO file (name, size, checksum, fileset, detection, detection_type, `timestamp`) VALUES ('{encode_punycode(file['name'])}', '{file['size']}', '{checksum}', @fileset_last, {detection}, '{detection_type}', NOW())"
+        query = f"INSERT INTO file (name, size, checksum, fileset, detection, detection_type, `timestamp`) VALUES ('{escape_string(encode_punycode(file['name']))}', '{file['size']}', '{checksum}', @fileset_last, {detection}, '{detection_type}', NOW())"
```
Use this pattern to have the library escape the variables:

```
        sql = "INSERT INTO `users` (`email`, `password`) VALUES (%s, %s)"
        cursor.execute(sql, ('webmaster@python.org', 'very-secret'))
        ```
        
        From: https://pymysql.readthedocs.io/en/latest/user/examples.html

ShivangNagta replied:
Yes, I was planning to convert all current queries to parametrised queries

rvanlaar commented on db_functions.py:
```diff
@@ -801,8 +800,7 @@ def populate_file(fileset, fileset_id, conn, detection):
             if punycode_need_encode(file['name']):
-                print(encode_punycode(file['name']))
-                query = f"INSERT INTO file (name, size, checksum, fileset, detection, detection_type, `timestamp`) VALUES ('{encode_punycode(file['name'])}', '{file['size']}', '{checksum}', @fileset_last, {detection}, '{detection_type}', NOW())"
+                query = f"INSERT INTO file (name, size, checksum, fileset, detection, detection_type, `timestamp`) VALUES ('{escape_string(encode_punycode(file['name']))}', '{file['size']}', '{checksum}', @fileset_last, {detection}, '{detection_type}', NOW())"
```
Same here, escape it by having PyMySQL escape it for you.

rvanlaar commented on requirements.txt:
```diff
@@ -1,10 +1,18 @@
 blinker==1.9.0
```
What's up with the requirements.txt changes in a commit for fixing punycode?

rvanlaar commented on tests/test_punycode.py:
```diff
@@ -0,0 +1,54 @@
+import pytest
```
No need to import pytest from here.

rvanlaar commented on tests/create/create_binary.py:
```diff
@@ -0,0 +1,66 @@
+
+output_dir = "../data/invalid_mac_binary/"
+
+with open(os.path.join(output_dir, "len_less_than_128_bytes.bin"), "wb") as f:
```
use `if __name__ == "__main__"` for executable python files.

rvanlaar commented on tests/test_compute_hash.py:
```diff
@@ -0,0 +1,24 @@
+# from compute_hash import (
```
Drop commented out code.

rvanlaar commented on tests/test_compute_hash.py:
```diff
@@ -0,0 +1,24 @@
+# )
+import sys
+import os
+sys.path.insert(0, ".")
```
Can you explain why this code piece is here?

ShivangNagta replied:
When running pytest from the main directory, it somehow doesn't add the current working directory to the path. So i had to add it manually in one of the test files.

rvanlaar commented on tests/test_compute_hash.py:
```diff
@@ -0,0 +1,24 @@
+import pytest
+from compute_hash import is_macbin
+
+def test_is_macbin():
```
Good job on writing tests.

rvanlaar commented on fileset.py:
```diff
@@ -50,11 +50,41 @@ def index():
+    finally:
+        conn.close()
+
+    return redirect(f'/')
```
No need to make an 'f' string out of a regular string.

rvanlaar commented on db_functions.py:
```diff
@@ -159,10 +161,23 @@ def insert_file(file, detection, src, conn):
+    columns.extend(['checksum', 'fileset', 'detection', 'detection_type', '`timestamp`'])
+    values.extend([f"'{checksum}'", '@fileset_last', str(detection), f"'{detection_type}'", 'NOW()'])
+
+    query = f"INSERT INTO file ({', '.join(columns)}) VALUES ({', '.join(values)})"
```
Add the values on execute.

rvanlaar commented on fileset.py:
```diff
@@ -50,11 +50,41 @@ def index():
+            conn.commit()
+            print("DATABASE CLEARED")
+    except Exception as e:
+        print(f"Error clearing database: {e}")
```
When you start refactoring the flask part more, use logging instead of print:  https://flask.palletsprojects.com/en/stable/logging/

rvanlaar commented on db_functions.py:
```diff
@@ -167,12 +167,21 @@ def insert_file(file, detection, src, conn):
     if extended_file_size:
-        columns.extend(['`size-r`', '`size-rd`'])
-        values.extend([f"'{file['size-r']}'", f"'{file['size-rd']}'"])
+        values.extend([f"'{file['size']}'", f"'{file['size-r']}'", f"'{file['size-rd']}'"])
```
What's going on here in this code? Please explain.

ShivangNagta replied:
Actually some dat files have 3 size entries (e.g. coming from the compute_hash.py utlity)  and some only 1 (e.g. scummvm.dat, which is the dumped data from scummvm detection entries), so I was trying to parse both the cases. But this logic has been improved in a later commit which I haven't made yet.

rvanlaar commented on db_functions.py:
```diff
@@ -203,9 +203,38 @@ def insert_filechecksum(file, checktype, conn):
         cursor.execute(query)
+        if "md5" not in checktype:
+            return
+        if (checktype[-1] == 'm' or checktype[-1] == 'd' or checktype[-1] == 'r'):
```
use `in`

rvanlaar commented:
Hi,

In general, nice work. This was more code then I'ld expect for one of the first PRs. Congrats. 

First a few general tips and tricks:
Install a pre-commit hook with ruff format: https://github.com/astral-sh/ruff-pre-commit
And format the codebase. 

If you're feeling brave also run Ruff's linter.

I've added comments next to the code. Please go through them and fix the issues.

One last and important part:  please check your unittests.

Regards, Roland.

ShivangNagta committed:
INTEGRITY: Additional error handling for raw rsrc and actual mac fork

ShivangNagta committed:
INTEGRITY: Add support for 7th mac variant - actual resource fork

ShivangNagta committed:
INTEGRITY: Update requirements.txt

ShivangNagta committed:
INTEGRITY: Fix punycode

* Match punycode with dumper companion
* Escape single quote in encoded punycode for sql query
* Add unit tests for punycode_need_encode and encode_punycode

ShivangNagta committed:
INTEGRITY: Add pre-commit hook for ruff.

Set up and ran ruff formatter and linter on the codebase.

sev- commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,9 @@
+repos:
```
Please remove this file. It should not be committed

sev- commented on clear.py:
```diff
@@ -7,26 +7,38 @@
 def truncate_all_tables(conn):
-    tables = ["filechecksum", "queue", "history", "transactions", "file", "fileset", "game", "engine", "log"]
+    tables = [
+        "filechecksum",
```
Please revert. This is an unnecessary format change

sev- commented on compute_hash.py:
```diff
@@ -3,55 +3,290 @@
 
 # CRC table
 CRC16_XMODEM_TABLE = [
-    0x0000, 0x1021, 0x2042, 0x3063, 0x4084, 0x50a5, 0x60c6, 0x70e7,
```
Please revert

sev- commented on clear.py:
```diff
@@ -7,26 +7,38 @@
 if __name__ == "__main__":
     base_dir = os.path.dirname(os.path.abspath(__file__))
-    config_path = os.path.join(base_dir, 'mysql_config.json')
+    config_path = os.path.join(base_dir, "mysql_config.json")
```
Why this syntax change is needed? Same below at line 56

sev- commented on compute_hash.py:
```diff
@@ -3,55 +3,290 @@
-        crc = ((crc << 8) & 0xff00) ^ CRC16_XMODEM_TABLE[(
-            (crc >> 8) & 0xff) ^ byte]
-    return crc & 0xffff
+        crc = ((crc << 8) & 0xFF00) ^ CRC16_XMODEM_TABLE[((crc >> 8) & 0xFF) ^ byte]
```
Please revert

sev- reviewed (changes requested):
Added couple of notes. It is very difficult to review with so many formatting changes.

I propose the following:

1. You re-run lint as you did on a freshly checked out repo
2. You create a new PR with only the formatting changes (and please avoid those I mearked)
3. I merge your formatiing changes, e.g. non-functional changes
4. You rebase this branch over the new code

This way, this PR will contain only the code changes, not the formatting

sev- commented on .gitignore:
```diff
@@ -2,4 +2,8 @@
-.pre-commit-config.yaml
\ No newline at end of file
+.pre-commit-config.yaml
+dumps
```
Please do not add these files to .gitignore. it is just your local thing and should not be in the project

sev- commented on .gitignore:
```diff
@@ -2,4 +2,8 @@
+dumps
+.pytest_cache
+mac_dats
+macresfork 2
```
These files also need to go away.

Also, please fix your editor so it adds a new line at the end of files.

sev- commented on compute_hash.py:
```diff
@@ -223,70 +247,91 @@ def appledouble_get_datafork(filepath):
-    with open(filepath, "rb") as file:
-        return create_checksum_pairs(checksum(file, alg, size, filepath), alg, size)
+        hashes = checksum(combined_forks, alg, size, filepath)
+        prefix = 'm'
```
I dropped `m:` checksums from scummvm.dat. So, you may remove this part

sev- commented on db_functions.py:
```diff
@@ -831,7 +911,7 @@ def find_matching_filesets(fileset, conn, status):
             matched_set = set()
             for key, value in file.items():
-                if key not in ["name", "size", "sha1", "crc"]:
+                if key not in ["name", "size", "size-r", "size-rd", "sha1", "crc"]:
```
Let's drop sha1 and crc from the DB. They have no use for us.

sev- commented on db_functions.py:
```diff
@@ -952,8 +1035,10 @@ def handle_matched_filesets(
 
 def delete_original_fileset(fileset_id, conn):
     with conn.cursor() as cursor:
+        print(fileset_id)
```
I think this is a leftover. You may consider introducing like 'debug mode' and print this information

sev- reviewed (commented):
Added my notes

ShivangNagta committed:
INTEGRITY: Add warning log and skip duplicate detection entry.

There are some entries in scummvm.dat that share the same megakey even with different gameid or title. During the initial seeding, these entries will now be skipped if one of them has already been added. Their engineid, gameid, platform, and language will be logged.

ShivangNagta committed:
INTEGRITY: Drop support for generating m prefix checksums for macfiles

ShivangNagta committed:
INTEGRITY: Remove sha1 and crc checksum addition to database

ShivangNagta committed:
INTEGRITY: Parse 'f' prefix checksum as normal front bytes checksum

ShivangNagta committed:
INTEGRITY: Rewrite set.dat processing logic.

ShivangNagta committed:
INTEGRITY: Handle multiple fileset references in log entries

ShivangNagta committed:
INTEGRITY: Drop set filesets if no candidates for matching - e.g mac files

ShivangNagta committed:
INTEGRITY: Add re-updation logic for set.dat

ShivangNagta committed:
INTEGRITY: Create extra indices in db for faster query

ShivangNagta committed:
INTEGRITY: Add additional log details for number of filesets in different categories

ShivangNagta committed:
INTEGRITY: Filter all candidates in descending order of matches instead of only the max one

ShivangNagta committed:
INTEGRITY: Remove set filesets with many to one mapping with a single fileset

ShivangNagta committed:
INTEGRITY: Stop processing for the fileset if it already exists - checked by key

ShivangNagta committed:
Skip certain logs while processing set.dat with skiplog flag

ShivangNagta committed:
INTEGRITY: Replace detection duplicate check by filename, size and checksum instead of megakey

ShivangNagta committed:
INTEGRITY: Improve filtering and navigation in fileset search

ShivangNagta committed:
INTEGRITY: Remove punycode encoding while loading to database, further convert \ to / in filepaths for filesystem independent parsing

ShivangNagta committed:
INTEGIRTY: Iteratively look for extra files if romof or cloneof field is present in the set.dat metadata. Filtering update.

ShivangNagta committed:
INTEGRITY: Add possible merge button inside filesets dashboard

ShivangNagta committed:
INTEGRITY: Filter files by filename instead of entire path, as detections do not necessarily store the entire filepath.

ShivangNagta committed:
INTEGRITY: Avoid adding duplicate files from detection entries

ShivangNagta committed:
INTEGRITY: Avoid detection file overriding on a match, when a similar file exist in a different directory. Add it as a normal non detection file.

ShivangNagta committed:
INTEGRITY: Create copy of the game data during lookup map creation to avoid issues due to mutability of python dictionaries.

sev- commented:
Merging as a milestone
